src/mlpack/methods/ann/rbm/binary_rbm_policy_impl.hpp

```diff
@@ -0,0 +1,159 @@
+/**
+ * @file binary_rbm.hpp
+ * @author Kris Singh
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_RBM_BINARY_RBM_POLICY_IMPL_HPP
+#define MLPACK_METHODS_ANN_RBM_BINARY_RBM_POLICY_IMPL_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+
+#include "binary_rbm_policy.hpp"
+
+namespace mlpack {
+namespace ann {
+
+template<typename DataType>
+BinaryRBMPolicy<DataType>
+    ::BinaryRBMPolicy(size_t visibleSize, size_t hiddenSize) :
+    visibleSize(visibleSize),
+    hiddenSize(hiddenSize)
+{
+  parameter.set_size((visibleSize * hiddenSize) + visibleSize + hiddenSize, 1);
+}
+
+// Reset function
+template<typename DataType>
+void BinaryRBMPolicy<DataType>::Reset()
+{
+  weight = DataType(parameter.memptr(), hiddenSize, visibleSize, false, false);
+  hiddenBias = DataType(parameter.memptr() + weight.n_elem,
+      hiddenSize, 1, false, false);
+  visibleBias = DataType(parameter.memptr() + weight.n_elem +
+      hiddenBias.n_elem , visibleSize, 1, false, false);
+}
+
+template<typename DataType>
+typename BinaryRBMPolicy<DataType>::ElemType BinaryRBMPolicy<DataType>
+    ::FreeEnergy(DataType&& input)
+{
+  HiddenPreActivation(std::move(input), std::move(preActivation));
+  SoftplusFunction::Fn(preActivation, preActivation);
```

Comment:
it might be faster to use ``arma::trunc_exp`` instead of checking using

```c++
if (x < DBL_MAX)
      return x > -DBL_MAX ? std::log(1 + std::exp(x)) : 0;
    return 1.0;
```

which performs a check for each value, the performance boost might be neglectable but I think it's worth a check.